OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+SimTK::Vec2 VandenBogert2011Muscle::calcFiberStaticEquilbirum(
+        double muscleLength, double activ) const {
+
+    // This code assumes that the muscle lengthing speed is 0.
+
+
+    //TODO: Code is not optimized.  Specifically the number of calls to
+    //calcImplicitResidual can be reduced
+
+    //TODO: Generalize with a Lambda function (will need help with that).
+    //TODO: calcImplicitResidual really only needs to calculate df_ds
+    //       (single element) for this function
+
+
+double tol=1e-8; //TODO : Look at changing to propotional of eps
+double a=0;
+double b=10;  //10 muscle lengths (more will slow convergence by adding steps)
+
+double x=(a+b)/2.0;
+double dx=2*tol;
+
+int neval=0;
+
+SimTK::Vec3 forceResAndDerivative;
+
+    //Perform a Newton Line Search
+while ((abs(dx)>=tol) && (neval<100)) {
+
+    neval++;
+
+    // Set a to be lower value and b to be upper value
+    a=min(a,b);
+    b=max(a,b);
+
+
+    forceResAndDerivative =
+            calcFiberStaticEquilibResidual(x, muscleLength, activ);
+    double fx = forceResAndDerivative[0];
+
+    //After the 1st iteration, use the new guess as a new upper or lower bound
+    if (neval>1) {
+        forceResAndDerivative =
+                calcFiberStaticEquilibResidual(a, muscleLength, activ);
+        double funcA = forceResAndDerivative[0];
+
+        if ((funcA *fx)>0){
+            a = x;}
+        else { b = x;}}
+
+    forceResAndDerivative =
+            calcFiberStaticEquilibResidual(x, muscleLength, activ);
+    double dfx= forceResAndDerivative[1];
+    //double forceRes=forceResAndDerivative[0];
+
+    dx =-fx/dfx;
+    double xdx=x-dx;
+
+    bool inInterval=((xdx>=a) && (xdx<=b));
+
+    forceResAndDerivative =
+            calcFiberStaticEquilibResidual(xdx, muscleLength, activ);
+    bool largeDeriv=abs(forceResAndDerivative[0])> (0.5*abs(fx));
+
+    if (~inInterval || largeDeriv) {
+        x=(a+b)/2;
+        dx=(a-b)/2;}
+    else
+        {x=xdx;};
+
+
+    };
+
+    // TODO:  Need to handle condition when number of loop iterations reaches
+    // neval limit. See Millard2012 Muscle Error Handling
+
+    SimTK::Vec2 vout;
+    vout[0]=x;   //projFiberLengthNorm
+    vout[1]=forceResAndDerivative[2];  // muscleForce
+    return vout;
+}
+
+
+
+
+//------------------------------------------------------------------------------
+//Calculate the ydot values to drive the residuals to 0 and "balance" the muscle
+SimTK::Vector VandenBogert2011Muscle::calcSolveMuscle(SimTK::State& s,
+         double activ, SimTK::Vector yDotInitialGuess) const {
+
+
+    ImplicitSystemDerivativeSolver sys(2, s, activ);
+
+    //TODO:  Need come up with reasonable bounds
+    SimTK::Vector lower_bounds(2);
+    lower_bounds[0] = -SimTK::Infinity;
+    lower_bounds[1] = -SimTK::Infinity;
+
+    SimTK::Vector upper_bounds(2);
+    upper_bounds[0] = SimTK::Infinity;
+    upper_bounds[1] = SimTK::Infinity;
```

Comment:
It should be straightforward to compute the maximum/minimum values of `adot`. And the maximum/minimum values of fiber velocity could maybe be taken from the `max_contraction_velocity` property (I don't remember the exact name).